app/src/main/kotlin/trees/AVLTree.kt

```diff
@@ -0,0 +1,110 @@
+        updateHeight(root)
+    }
+
+    private fun recursiveAdd(currentNode: AVLNode<K, V>?, node: AVLNode<K, V>): AVLNode<K, V> {
+        if (currentNode == null) return node
+        if (currentNode.key < node.key) {
+            currentNode.right = recursiveAdd(currentNode.right, node)
+        }
+        if (currentNode.key > node.key) {
+            currentNode.left = recursiveAdd(currentNode.left, node)
+        }
+        if (currentNode.key == node.key) {
+            return node
+        }
+        return balance(currentNode)
+    }
+
+    override fun remove(key: K) {
+        val node = find(key) ?: throw IllegalStateException("There is no such node with key $key in the tree")
+        root = recursiveRemove(root, node)
+        updateHeight(root)
+    }
+
+    private fun recursiveRemove(currentNode: AVLNode<K, V>?, node: AVLNode<K, V>): AVLNode<K, V>? {
+        if (currentNode == null) return null
+        if (currentNode.key < node.key) {
+            currentNode.right = recursiveRemove(currentNode.right, node)
+        } else if (currentNode.key > node.key) {
+            currentNode.left = recursiveRemove(currentNode.left, node)
+        } else {
+            if (currentNode.left == null || currentNode.right == null)
+                return currentNode.left ?: currentNode.right
+            val minRightSubtree = getMinNode(currentNode.right!!)
+            currentNode.key = minRightSubtree.key
+            currentNode.value = minRightSubtree.value
+            currentNode.right = recursiveRemove(currentNode.right, minRightSubtree)
+        }
+        return balance(currentNode)
+    }
+
+
+    override fun balance(node: AVLNode<K, V>, afterRemove: Boolean): AVLNode<K, V> {
+        updateHeight(node)
+        when (getBalanceFactor(node)) {
+            -2 -> {
+                if (getBalanceFactor(node.left) == 1) {
+                    node.left = node.left?.let { rotateLeft(it) }
+                    updateHeight(node.left)
+                }
+                val balancedNode = rotateRight(node)
+                updateHeight(balancedNode)
+                return balancedNode
+            }
+
+            2 -> {
+                if (getBalanceFactor(node.right) == -1) {
+                    node.right = node.right?.let { rotateRight(it) }
+                    updateHeight(node.right)
+                }
+                val balancedNode = rotateLeft(node)
+                updateHeight(balancedNode)
+                return balancedNode
+            }
+        }
+        return node
+    }
+
+    internal fun getHeight(node: AVLNode<K, V>?): Int {
+        return node?.height ?: -1
+    }
+
+    private fun getBalanceFactor(node: AVLNode<K, V>?): Int {
+        if (node == null) return 0
+        return getHeight(node.right) - getHeight(node.left)
+    }
+
+    private fun updateHeight(node: AVLNode<K, V>?) {
+        node?.let { it.height = max(getHeight(it.left), getHeight(it.right)) + 1 }
```

Comment:
В Котлине несколько [скоуп функций](https://kotlinlang.org/docs/scope-functions.html), каждую удобно применять в соответствующих ситуациях:
```suggestion
        node?.run { height = max(getHeight(left), getHeight(right)) + 1 }
```

Далее не буду подобные места комментировать, но скопу функции изучить полезно для повышения компактности и читаемости кода.